fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/DefaultResourcePackMixin.java

```diff
@@ -19,55 +19,110 @@
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.JarURLConnection;
 import java.net.URL;
-import java.util.Enumeration;
+import java.net.URLConnection;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.function.Predicate;
 
+import org.jetbrains.annotations.Nullable;
 import org.spongepowered.asm.mixin.Mixin;
-import org.spongepowered.asm.mixin.injection.At;
-import org.spongepowered.asm.mixin.injection.Inject;
-import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+import org.spongepowered.asm.mixin.Overwrite;
+import org.spongepowered.asm.mixin.Unique;
 
 import net.minecraft.resource.DefaultResourcePack;
 import net.minecraft.resource.DirectoryResourcePack;
+import net.minecraft.resource.ResourcePack;
 import net.minecraft.resource.ResourceType;
+import net.minecraft.resource.ZipResourcePack;
 import net.minecraft.util.Identifier;
 
+/**
+ * Make the default resource pack use the MC jar directly instead of the full classpath.
+ * This is a major speed improvement, as well as a bugfix (it prevents other mod jars from overriding MC's resources).
+ */
 @Mixin(DefaultResourcePack.class)
-public class DefaultResourcePackMixin {
-	@Inject(method = "findInputStream", at = @At("HEAD"), cancellable = true)
-	protected void onFindInputStream(ResourceType resourceType, Identifier identifier, CallbackInfoReturnable<InputStream> callback) {
-		if (DefaultResourcePack.resourcePath != null) {
-			// Fall through to Vanilla logic, they have a special case here.
-			return;
-		}
+public abstract class DefaultResourcePackMixin {
+	/**
+	 * Redirect all resource access to the MC jar zip pack.
+	 */
+	final ResourcePack fabric_mcJarPack = createJarZipPack();
 
-		String path = resourceType.getDirectory() + "/" + identifier.getNamespace() + "/" + identifier.getPath();
-		URL found = null;
+	@Unique
+	private ResourcePack createJarZipPack() {
+		ResourceType type;
 
+		if (getClass().equals(DefaultResourcePack.class)) {
+			// Server pack
+			type = ResourceType.SERVER_DATA;
+		} else {
+			// Client pack
+			type = ResourceType.CLIENT_RESOURCES;
+		}
+
+		// Locate MC jar by finding the URL that contains the assets root.
 		try {
-			Enumeration<URL> candidates = DefaultResourcePack.class.getClassLoader().getResources(path);
+			URL assetsRootUrl = DefaultResourcePack.class.getResource("/" + type.getDirectory() + "/.mcassetsroot");
+			URLConnection connection = assetsRootUrl.openConnection();
```

Comment:
Need to close the connection after you exit this method to prevent resource leak

Also override/overwrite `close` to close this backing resource pack as well.

Reply:
No, it's not closeable.
> Invoking the close() methods on the InputStream or OutputStream of an URLConnection after a request may free network resources associated with this instance, unless particular protocol specifications specify different behaviours for it.

Reply:
> Also override/overwrite `close` to close this backing resource pack as well.

Well: :smile:

>
```java
 	/**
	 * @author FabricMC
	 * @reason Close the resource pack we redirect resource access to.
	 */
	@Overwrite
	public void close() {
		fabric_mcJarPack.close();
	}
```